crates/rpc-types-engine/src/payload/mod.rs

```diff
@@ -23,6 +25,269 @@ pub enum OpExecutionPayload {
+                    fee_recipient: fee_recipient
+                        .ok_or_else(|| serde::de::Error::missing_field("feeRecipient"))?,
+                    state_root: state_root
+                        .ok_or_else(|| serde::de::Error::missing_field("stateRoot"))?,
+                    receipts_root: receipts_root
+                        .ok_or_else(|| serde::de::Error::missing_field("receiptsRoot"))?,
+                    logs_bloom: logs_bloom
+                        .ok_or_else(|| serde::de::Error::missing_field("logsBloom"))?,
+                    prev_randao: prev_randao
+                        .ok_or_else(|| serde::de::Error::missing_field("prevRandao"))?,
+                    block_number: block_number
+                        .ok_or_else(|| serde::de::Error::missing_field("blockNumber"))?,
+                    gas_limit: gas_limit
+                        .ok_or_else(|| serde::de::Error::missing_field("gasLimit"))?,
+                    gas_used: gas_used.ok_or_else(|| serde::de::Error::missing_field("gasUsed"))?,
+                    timestamp: timestamp
+                        .ok_or_else(|| serde::de::Error::missing_field("timestamp"))?,
+                    extra_data: extra_data
+                        .ok_or_else(|| serde::de::Error::missing_field("extraData"))?,
+                    base_fee_per_gas: base_fee_per_gas
+                        .ok_or_else(|| serde::de::Error::missing_field("baseFeePerGas"))?,
+                    block_hash: block_hash
+                        .ok_or_else(|| serde::de::Error::missing_field("blockHash"))?,
+                    transactions: transactions
+                        .ok_or_else(|| serde::de::Error::missing_field("transactions"))?,
+                };
+
+                // return V4 is withdrawals_root is present
+                if let Some(withdrawals_root) = withdrawals_root {
+                    if let (Some(blob_gas_used), Some(excess_blob_gas)) =
+                        (blob_gas_used, excess_blob_gas)
+                    {
+                        return Ok(OpExecutionPayload::V4(OpExecutionPayloadV4 {
+                            payload_inner: ExecutionPayloadV3 {
+                                payload_inner: ExecutionPayloadV2 {
+                                    payload_inner: v1,
+                                    withdrawals: withdrawals.unwrap(),
+                                },
+                                blob_gas_used,
+                                excess_blob_gas,
+                            },
+                            withdrawals_root,
+                        }));
+                    };
+                // return V3 is withdrawals_root is absent, but blob_gas_used and excess_blob_gas is present
+                } else if let (Some(blob_gas_used), Some(excess_blob_gas)) =
+                    (blob_gas_used, excess_blob_gas)
+                {
+                    return Ok(OpExecutionPayload::V3(ExecutionPayloadV3 {
```

Comment:
I think this can be simplified because we create the v3 here twice now and have the some blob gas checks,

we can do those first and only then decide if we need v4 or v3 based on the withdrawalsroot